### PR TITLE
Enable nullables part 1

### PR DESCRIPTION
**Pull request by SimonCropp**
this is a first pass at enabling nullables. 

I think this is a big enough chunk to do a code review

### CODE REVIEW

**Comment by SimonCropp** [comment]
note i have already enable nullables for the tests

File: src/Polly/AsyncPolicy.TResult.cs [hunk 1]
```diff
@@ -1,4 +1,5 @@
-﻿namespace Polly;
+﻿#nullable enable
```
Comment (martincostello) [inline src/Polly/AsyncPolicy.TResult.cs:1]: Is the plan to remove these and just set it in the project file once it's "done"?

File: src/Polly/AsyncPolicy.NonGenericImplementation.cs [hunk 2]
```diff
@@ -40,5 +40,6 @@ public abstract partial class AsyncPolicy
         Context context,
         CancellationToken cancellationToken,
         bool continueOnCapturedContext
-    );
+    )
+        where TResult : notnull;
```
Comment (martincostello) [inline src/Polly/AsyncPolicy.NonGenericImplementation.cs:44]: Does this preclude someone from a return type like `Task<Foo?>`?

File: src/Polly/Caching/AsyncCacheEngine.cs [hunk 3]
```diff
@@ -43,7 +45,7 @@ internal static class AsyncCacheEngine
         if (cacheHit)
```
Comment (martincostello) [inline src/Polly/Caching/AsyncCacheEngine.cs:45]: Is there a way to annotate it so the compiler knows that if `cacheHit` is `true` then `valueFromCache` is non-null?

Comment (martincostello) [reply src/Polly/Caching/AsyncCacheEngine.cs:45]: Or, are there use cases where `null` is a valid value in a cache (e.g. _"I've looked this up and it doesn't exist, so cache that fact"_), and we should therefore allow null?

File: src/Polly/Caching/AsyncGenericCacheProvider.cs [hunk 4]
```diff
@@ -9,16 +10,17 @@ namespace Polly.Caching;
...
 internal class AsyncGenericCacheProvider<TCacheFormat> : IAsyncCacheProvider<TCacheFormat>
+    where TCacheFormat : notnull
 {
     private readonly IAsyncCacheProvider _wrappedCacheProvider;
 
     internal AsyncGenericCacheProvider(IAsyncCacheProvider nonGenericCacheProvider)
         => _wrappedCacheProvider = nonGenericCacheProvider ?? throw new ArgumentNullException(nameof(nonGenericCacheProvider));
 
-    async Task<(bool, TCacheFormat)> IAsyncCacheProvider<TCacheFormat>.TryGetAsync(string key, CancellationToken cancellationToken, bool continueOnCapturedContext)
+    async Task<(bool, TCacheFormat?)> IAsyncCacheProvider<TCacheFormat>.TryGetAsync(string key, CancellationToken cancellationToken, bool continueOnCapturedContext)
```
Comment (martincostello) [inline src/Polly/Caching/AsyncGenericCacheProvider.cs:20]: A named tuple could be good here as this is internal.

File: src/Polly/Caching/AsyncSerializingCacheProvider.cs [hunk 5]
```diff
@@ -36,10 +38,10 @@ public AsyncSerializingCacheProvider(IAsyncCacheProvider<TSerialized> wrappedCac
     /// A <see cref="Task{TResult}" /> promising as Result a tuple whose first element is a value indicating whether
     /// the key was found in the cache, and whose second element is the value from the cache (null if not found).
     /// </returns>
-    public async Task<(bool, object)> TryGetAsync(string key, CancellationToken cancellationToken, bool continueOnCapturedContext)
+    public async Task<(bool, object?)> TryGetAsync(string key, CancellationToken cancellationToken, bool continueOnCapturedContext)
```
Comment (martincostello) [inline src/Polly/Caching/AsyncSerializingCacheProvider.cs:41]: +1 to named tuple.

Comment (martincostello) [reply src/Polly/Caching/AsyncSerializingCacheProvider.cs:41]: Nevermind, I see that it's implementing a public interface.

File: src/Polly/CircuitBreaker/RollingHealthMetrics.cs [hunk 6]
```diff
@@ -24,14 +25,14 @@ public void IncrementSuccess_NeedsLock()
     {
         ActualiseCurrentMetric_NeedsLock();
 
-        _currentWindow.Successes++;
+        _currentWindow!.Successes++;
```
Comment (martincostello) [inline src/Polly/CircuitBreaker/RollingHealthMetrics.cs:28]: Is there an annotation we can add to hint the compiler that calls to a certain method will result in this being non-null?

File: src/Polly/CircuitBreaker/SingleHealthMetrics.cs [hunk 7]
```diff
@@ -7,22 +8,22 @@ internal class SingleHealthMetrics : IHealthMetrics
...
+    private HealthCount? _current;
 
     public SingleHealthMetrics(TimeSpan samplingDuration) => _samplingDuration = samplingDuration.Ticks;
 
     public void IncrementSuccess_NeedsLock()
     {
         ActualiseCurrentMetric_NeedsLock();
 
-        _current.Successes++;
+        _current!.Successes++;
```
Comment (martincostello) [inline src/Polly/CircuitBreaker/SingleHealthMetrics.cs:19]: Same question.

**Review by martincostello — COMMENTED** [review]
Maybe I've just gotten confused between _nullable_ and _nullability_, but I'm not sure if the constraints for `TResult: nonnull` are correct.

File: src/Polly/AsyncPolicy.TResult.cs (hunk 1 above)
Comment (SimonCropp) [reply src/Polly/AsyncPolicy.TResult.cs:1]: yep. `#nullable enable` allows us to gradually transition

File: src/Polly/AsyncPolicy.NonGenericImplementation.cs (hunk 2 above)
Comment (SimonCropp) [reply src/Polly/AsyncPolicy.NonGenericImplementation.cs:44]: yes. is that correct?

Comment (martincostello) [reply src/Polly/AsyncPolicy.NonGenericImplementation.cs:44]: That doesn't sound right to me - if a user wants to return `null` from their policy they should be able to.

**Comment by SimonCropp** [comment]
@martincostello i think in retrospect this one is too large to reason about. i will submit a few smaller ones

**Pull request closed without merging**